src/octopath.cpp

```diff
@@ -106,199 +153,333 @@ bool OctoClass::OctoRRG(const Eigen::Vector3d &p0,
+      path->insert(path->begin(), pos);
+      while (current_key != initial_key) {
+        if (!indexed_free_keys.Key2Index(current_key, &current_index)) {
+          ROS_INFO("[mapper] Astar failed when retrieving path from initial to final point!");
+          return false;
         }
-    } else {
-        ROS_WARN("[mapper] RRG Error: Could not find a path from p0 to pf!");
-        *n_rrg_nodes = n_nodes;
-        *plan_time = (ros::Time::now() - t0).toSec();
-        return false;
+        // ROS_INFO("Come from: %d", int(current_index));
+        current_key = come_from[current_index];
+        current_pos = tree_inflated_.keyToCoord(current_key);
+        pos = Eigen::Vector3d(current_pos.x(), current_pos.y(), current_pos.z());
+        path->insert(path->begin(), pos);
+      }
+      break;
     }
 
-    // Prune results if requested
-    if (prune_result) {
-        this->PathPruning(sol_path, free_space_only, path);
-    } else {
-        *path = sol_path;
+    // Check neighbors
+    std::vector<octomap::OcTreeKey> neighbor_keys;
+    this->GetNodeNeighbors(current_key, node_sizes[current_index], &neighbor_keys);
+    const uint n_neighbors = neighbor_keys.size();
+    // ROS_INFO("Current index: %d (%f, %f, %f) has %d neighbors", int(current_index),
+    //           current_pos.x(), current_pos.y(), current_pos.z(), int(n_neighbors));
+
+    for (uint i = 0; i < n_neighbors; i++) {
+      // Get neighbor index
+      uint neighbor_index;
+      if (!indexed_free_keys.Key2Index(neighbor_keys[i], &neighbor_index)) {
+        ROS_INFO("[mapper] Astar failed: Error retrieving index for neighbor node!");
+        return false;
+      }
+
+      // Compute an added cost based on the distance between the neighbor and the nearest obstacle
+      const octomap::point3d neighbor_pos = tree_inflated_.keyToCoord(neighbor_keys[i]);
+      if (node_distance_to_wall[neighbor_index] < 0.0) {
+        node_distance_to_wall[neighbor_index] = this->NearestObstaclePathCost(neighbor_pos);
+      }
+
+      // Check whether this neighbor leads to a new path
+      const double tentative_cost = cost_so_far[current_index] + (current_pos - neighbor_pos).norm()
+                                                               + node_distance_to_wall[neighbor_index];
+      if (tentative_cost < cost_so_far[neighbor_index]) {
+        // Fill up A* information for neighbor
+        cost_so_far[neighbor_index] = tentative_cost;
+        const double priority = tentative_cost + (neighbor_pos - pf).norm();
+        queue.put(neighbor_keys[i], priority);
+        come_from[neighbor_index] = current_key;
+      }
     }
-
-    // Populate time and nodes
-    *n_rrg_nodes = n_nodes;
-    *plan_time = (ros::Time::now() - t0).toSec();
-
-    // ROS_INFO("[mapper] nNodes: %d", static_cast<int>(obj_rrg.rrgraph_.n_nodes_));
-    // ROS_INFO("[mapper] nEdges: %d", static_cast<int>(obj_rrg.rrgraph_.n_edges_));
-
-    return true;
+  }
```

Comment:
I have ideas on how to optimize this...but I'm withholding them since I just don't think the optimizations are worth the dev time right now

Reply:
I'm really curious on how you'd optimize this portion. Should I leave a comment in code about these ideas?

Reply:
yeah I just thought of one optimization that is pretty simple - if the cost before obstacle path cost is already greater than cost_so_far[neighbor_index], then there's no need to call NearestObstaclePathCost since it's always >0 (this seems like a sort of expensive function)

There may be others!

Reply:
I just implemented your idea, and here are the results for a test run I made:
- Before making that check, the function `NearestObstaclePathCost` was called `6939` times
- After making that check and skipping the neighbor if cost before obstacle path cost is already greater than cost_so_far[neighbor_index], the function `NearestObstaclePathCost` was called `6938` times

Not so encouraging, but I see it as a good optimization. I'll leave it in as you suggested